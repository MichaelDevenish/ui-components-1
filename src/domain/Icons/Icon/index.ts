export { Icon, IIconProps } from './Icon'
export {
  IconType,
  FontAwesomeIconType,
  FontAwesomeIconTypeNoPrefix,
  IntelliIconType,
  IntelliIconTypeNoPrefix
} from './types'
