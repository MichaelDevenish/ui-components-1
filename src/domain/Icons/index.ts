export * from './FontAwesomeIcon'
export * from './IntelliIcon'
export * from './Icon'
