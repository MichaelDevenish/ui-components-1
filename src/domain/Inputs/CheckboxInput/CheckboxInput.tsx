import React from 'react'
import { IGenericInputProps, Input } from '../Input'
import classNames from 'classnames'

const style = require('./style.scss')

export interface ICheckboxInputProps extends IGenericInputProps {
  /** Label to display next to the checkbox */
  label: JSX.Element | string
  /** If true, the checkbox is wrapped by a button */
  isButton?: boolean
}

export class CheckboxInput extends React.PureComponent<ICheckboxInputProps> {
  public static defaultProps: Partial<ICheckboxInputProps> = {
    isButton: false
  }

  get infoLabel (): JSX.Element | null {
    const {
      name,
      label,
      isButton
    } = this.props

    if (!label) {
      return null
    }

    return (
      <label
        htmlFor={name}
        className={classNames('checkbox', { 'checkbox-button': isButton })}
      >
        {label}
      </label>
    )
  }

  public render (): JSX.Element {
    const {
      value,
      className,
      ...props
    } = this.props

    return (
      <div
        className={classNames('checkbox-input', style.checkboxInput, className)}
      >
        <Input
          {...props}
          isChecked={Boolean(value)}
          type='checkbox'
        />
        {this.infoLabel}
      </div>
    )
  }
}
