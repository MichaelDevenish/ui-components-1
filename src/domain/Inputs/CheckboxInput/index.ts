export { CheckboxInput, ICheckboxInputProps } from './CheckboxInput'
