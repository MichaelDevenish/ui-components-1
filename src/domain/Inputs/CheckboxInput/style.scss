@import '../../../common/sass/variables';

:local(.checkboxInput) {

  $icon-width: 20px;

  input {
    height: 0;
    opacity: 0;
    position: absolute;
    width: 0;

    & + .checkbox {
      width: 100%;
    }

    &:disabled + .checkbox {
      color: $n500;
      cursor: not-allowed;

      &::before {
        background-color: $n150;
        border-color: $n300;
      }
    }

    &:focus + label {
      &.checkbox {
        border-color: $i400;

        &::before {
          background-color: $n300;
          border-color: $i400;
        }
      }

      &.checkbox-button {
        background-color: $n200;
        border-color: $i400;
      }
    }

    &:checked + .checkbox {
      border-color: $i600;

      &::before {
        border-color: $i400;
      }

      &::after {
        opacity: 1;
      }

      &:hover {
        border-color: $i400;

        &::before {
          border-color: $i400;
        }
      }
    }

    &:checked + .checkbox-button {
      background-color: $i100;
      color: $i600;

      &:hover {
        background-color: $i100;

        &::before {
          background-color: $i100;
        }
      }
    }

    &:disabled + .checkbox-button {
      background-color: $n150;
      color: $n500;
      cursor: not-allowed;

      &::before {
        background-color: $n150;
        border-color: $n300;
      }
    }
  }

  .checkbox {
    align-self: stretch;
    color: $n800;
    cursor: pointer;
    font-size: $fz-small;
    line-height: $lh-small;
    margin-left: 0;
    margin-right: 0;
    min-height: calc(#{$spacing-2xsmall} * 2 + #{$icon-width});
    padding: $spacing-2xsmall $spacing-medium $spacing-2xsmall $spacing-xlarge;
    position: relative;
    transition: .25s ease-out, color .25s ease-out;

    &::before {
      background-color: $n150;
      border: 2px solid $n400;
      border-radius: $border-radius;
      content: '';
      display: inline-block;
      height: $icon-width;
      left: 0;
      position: absolute;
      top: 4px;
      transition: .25s ease-out, color .25s ease-out;
      vertical-align: top;
      width: $icon-width;
    }

    &::after {
      background: $i400;
      border-radius: $border-radius;
      box-shadow:
      2px 0 0 $i400,
      4px 0 0 $i400,
      4px -2px 0 $i400,
      4px -4px 0 $i400,
      4px -6px 0 $i400,
      4px -8px 0 $i400;
      content: '';
      height: 3px;
      left: 4px;
      opacity: 0;
      position: absolute;
      top: 13px;
      transform: rotate(45deg);
      transition: .25s ease-out;
      width: 3px;
    }

    &.checkbox-button {
      padding: $spacing-small $spacing-large $spacing-small ($icon-width + $spacing-large);

      &::before {
        left: 12px;
        top: 12px;
      }

      &::after {
        left: 16px;
        top: 21px;
      }
    }

    &:hover {

      &::before {
        background-color: $n300;
        border-color: $n600;
      }
    }
  }

  .checkbox-button {
    background-color: $n100;
    border: 1px solid $n300;
    border-radius: $border-radius;
    min-width: 240px;

    &:hover {
      background-color: $n200;
    }
  }
}
