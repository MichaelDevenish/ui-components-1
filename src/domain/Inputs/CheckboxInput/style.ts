import styled from 'styled-components'

import { Props } from '../../../common'
import { styleForMargins } from '../../Spacers/services/margins'

export interface IStyledCheckboxInputWrapperProps {
  margins?: Props.IMargins
}

const StyledCheckboxInput = styled.input`
  line-height: 16px;

  margin: 0;

  &::-ms-clear {
    display: none;
  }
`

const StyledCheckboxInputWrapper = styled.div`
  ${(props: IStyledCheckboxInputWrapperProps) => styleForMargins(props.margins)}
`

export {
  StyledCheckboxInput,
  StyledCheckboxInputWrapper
}
