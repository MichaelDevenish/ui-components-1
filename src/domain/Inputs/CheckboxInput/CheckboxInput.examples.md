#### Basic Input

```jsx
initialState = { value: false, value2: false };

<div>
  <CheckboxInput
    name='example-checkbox'
    label='This is a checkbox input'
    value={state.value}
    handleChange={() => setState({value: !state.value})}
  />
  <CheckboxInput
    isDisabled
    name='example-disabled-checkbox'
    label='This is a disabled checkbox input'
    value={state.value2}
    handleChange={() => setState({value2: !state.value2})}
  />
</div>
```

#### Input Button

```jsx
initialState = { button: false, button2: false };

<div>
  <CheckboxInput
    isButton
    name='button-checkbox'
    label='This is a checkbox input button'
    value={state.button}
    handleChange={() => setState({button: !state.button})}
  />
  <CheckboxInput
    isButton
    isDisabled
    name='button-disabled-checkbox'
    label='This is a disabled checkbox input button'
    value={state.button2}
    handleChange={() => setState({button2: !state.button2})}
  />
</div>
```

#### Input with JSX

```jsx
import { FontAwesomeIcon } from '@Domain/Icons';

initialState = { star: false };

<CheckboxInput
  name='star-checkbox'
  label={<div><FontAwesomeIcon type='solid' icon='star' /> I am a star</div>}
  value={state.star}
  handleChange={() => setState({star: !state.star})}
/>
```
