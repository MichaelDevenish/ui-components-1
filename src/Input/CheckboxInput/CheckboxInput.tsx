import React from 'react'
import { InfoLabel } from './style'
import { GenericInputProps, Input } from '../Input'
import classNames from 'classnames'

export interface CheckboxProps extends GenericInputProps {
  /** Label to display next to the checkbox */
  label?: JSX.Element | string
}

export class CheckboxInput extends React.PureComponent<CheckboxProps> {
  get infoLabel (): JSX.Element | null {
    const {
      name,
      label
    } = this.props

    if (!label) {
      return null
    }

    return (
      <InfoLabel
        htmlFor={name}
      >
        {label}
      </InfoLabel>
    )
  }

  public render (): JSX.Element {
    const {
      value,
      className
    } = this.props

    return (
      <div
        className={classNames('checkbox-input', className)}
      >
        <Input
          {...this.props}
          className={undefined}
          isChecked={Boolean(value) || undefined}
          type='checkbox'
        />
        {this.infoLabel}
      </div>
    )
  }
}
