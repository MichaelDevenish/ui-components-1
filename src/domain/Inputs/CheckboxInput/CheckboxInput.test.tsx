import { shallow } from 'enzyme'
import React from 'react'

import { CheckboxInput } from './CheckboxInput'
import { FontAwesomeIcon } from '@Domain/Icons'

describe('<CheckboxInput />', () => {
  it(`should render a checkbox input without button`, () => {
    const wrapper = shallow(
      <CheckboxInput
        name='test-input'
        label='Hey I am a test string'
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a checkbox input with button`, () => {
    const wrapper = shallow(
      <CheckboxInput
        name='test-button-input'
        label='Hey I am a test string'
        isButton
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a checkbox input with a component as the label`, () => {
    const wrapper = shallow(
      <CheckboxInput
        name='test-jsx-input'
        label={<FontAwesomeIcon type='star' />}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
