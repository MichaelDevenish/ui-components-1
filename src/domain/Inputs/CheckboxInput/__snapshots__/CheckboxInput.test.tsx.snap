// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxInput /> should render a checkbox input with a component as the label 1`] = `
<styled.div>
  <div
    className="checkbox-input checkboxInput"
  >
    <styled.input
      checked={false}
      className="input "
      data-component-type="checkbox_input"
      disabled={false}
      id="test-jsx-input"
      name="test-jsx-input"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="checkbox"
      htmlFor="test-jsx-input"
    >
      <Memo(FontAwesomeIcon)
        icon="star"
        type="solid"
      />
    </label>
  </div>
</styled.div>
`;

exports[`<CheckboxInput /> should render a checkbox input with button 1`] = `
<styled.div>
  <div
    className="checkbox-input checkboxInput"
  >
    <styled.input
      checked={false}
      className="input "
      data-component-type="checkbox_input"
      disabled={false}
      id="test-button-input"
      name="test-button-input"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="checkbox checkbox-button"
      htmlFor="test-button-input"
    >
      Hey I am a test string
    </label>
  </div>
</styled.div>
`;

exports[`<CheckboxInput /> should render a checkbox input without button 1`] = `
<styled.div>
  <div
    className="checkbox-input checkboxInput"
  >
    <styled.input
      checked={false}
      className="input "
      data-component-type="checkbox_input"
      disabled={false}
      id="test-input"
      name="test-input"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="checkbox"
      htmlFor="test-input"
    >
      Hey I am a test string
    </label>
  </div>
</styled.div>
`;

exports[`<CheckboxInput /> should render a checked checkbox input without button 1`] = `
<styled.div>
  <div
    className="checkbox-input checkboxInput"
  >
    <styled.input
      checked={true}
      className="input "
      data-component-type="checkbox_input"
      disabled={false}
      id="test-input"
      name="test-input"
      onChange={[Function]}
      type="checkbox"
      value="true"
    />
    <label
      className="checkbox"
      htmlFor="test-input"
    >
      Hey I am a test string
    </label>
  </div>
</styled.div>
`;
